File: lab/api/models/video_source_intake.py
"""Public contracts for project-scoped multimodal video source intake.

The API deliberately exposes domain identifiers and short-lived upload/delivery
instructions only. Provider configuration and durable storage locators remain
server-side.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, HttpUrl, field_validator


class StrictRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class SourceType(str, Enum):
    BINARY_VIDEO = "binary_video"
    BINARY_IMAGE = "binary_image"
    BINARY_AUDIO = "binary_audio"
    PUBLIC_LINK = "public_link"
    PASTED_TEXT = "pasted_text"


class SourceStatus(str, Enum):
    PENDING_VALIDATION = "pending_validation"
    PROCESSING = "processing"
    READY = "ready"
    METADATA_UNAVAILABLE = "metadata_unavailable"
    FAILED = "failed"
    REPLACEMENT_PENDING = "replacement_pending"
    SUPERSEDED = "superseded"
    REMOVED = "removed"
    ORPHAN_CLEANUP_NEEDED = "orphan_cleanup_needed"


class FolderStatus(str, Enum):
    COLLECTING = "collecting"
    READY = "ready"
    CHANGED_AFTER_READY = "changed_after_ready"
    ARCHIVED = "archived"


class EnqueueStatus(str, Enum):
    NOT_REQUESTED = "not_requested"
    ENQUEUE_PENDING = "enqueue_pending"
    ENQUEUED = "enqueued"
    ENQUEUE_FAILED = "enqueue_failed"


class OpenVideoSourceFolderRequest(StrictRequest):
    project_id: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("projectId", "project_id"),
        serialization_alias="projectId",
    )
    content_id: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("contentId", "content_id"),
        serialization_alias="contentId",
    )


class AddTextRequest(StrictRequest):
    text: str = Field(..., min_length=1, max_length=100_000)
    idempotency_key: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("idempotencyKey", "idempotency_key"),
        serialization_alias="idempotencyKey",
    )
    expected_revision: int = Field(
        ..., ge=0,
        validation_alias=AliasChoices("expectedRevision", "expected_revision"),
        serialization_alias="expectedRevision",
    )


class AddLinkRequest(StrictRequest):
    url: HttpUrl = Field(..., max_length=2048)
    idempotency_key: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("idempotencyKey", "idempotency_key"),
        serialization_alias="idempotencyKey",
    )
    expected_revision: int = Field(
        ..., ge=0,
        validation_alias=AliasChoices("expectedRevision", "expected_revision"),
        serialization_alias="expectedRevision",
    )


class CreateUploadSessionRequest(StrictRequest):
    source_type: Literal["binary_video", "binary_image", "binary_audio"] = Field(
        ...,
        validation_alias=AliasChoices("sourceType", "source_type"),
        serialization_alias="sourceType",
    )
    file_name: str = Field(
        ..., min_length=1, max_length=255,
        validation_alias=AliasChoices("fileName", "file_name"),
        serialization_alias="fileName",
    )
    mime_type: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("mimeType", "mime_type"),
        serialization_alias="mimeType",
    )
    byte_size: int = Field(
        ..., gt=0, le=200 * 1024 * 1024,
        validation_alias=AliasChoices("byteSize", "byte_size"),
        serialization_alias="byteSize",
    )
    checksum_sha256: str = Field(
        ..., pattern=r"^[a-fA-F0-9]{64}$",
        validation_alias=AliasChoices("checksumSha256", "checksum_sha256"),
        serialization_alias="checksumSha256",
    )
    expected_revision: int = Field(
        ..., ge=0,
        validation_alias=AliasChoices("expectedRevision", "expected_revision"),
        serialization_alias="expectedRevision",
    )
    idempotency_key: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("idempotencyKey", "idempotency_key"),
        serialization_alias="idempotencyKey",
    )
    replace_source_id: str | None = Field(
        default=None, min_length=1, max_length=128,
        validation_alias=AliasChoices("replaceSourceId", "replace_source_id"),
        serialization_alias="replaceSourceId",
    )

    @field_validator("file_name")
    @classmethod
    def clean_file_name(cls, value: str) -> str:
        cleaned = value.strip().replace("\\", "_").replace("/", "_")
        if cleaned in {"", ".", ".."}:
            raise ValueError("Invalid file name")
        return cleaned


class UploadedPartRequest(StrictRequest):
    part_number: int = Field(
        ..., ge=1, le=10_000,
        validation_alias=AliasChoices("partNumber", "part_number"),
        serialization_alias="partNumber",
    )
    etag: str = Field(..., min_length=1, max_length=256)
    checksum_sha256: str = Field(
        ...,
        pattern=r"^[a-fA-F0-9]{64}$",
        validation_alias=AliasChoices("checksumSha256", "checksum_sha256"),
        serialization_alias="checksumSha256",
    )
    size_bytes: int = Field(
        ..., gt=0, le=64 * 1024 * 1024,
        validation_alias=AliasChoices("sizeBytes", "size_bytes"),
        serialization_alias="sizeBytes",
    )


class CompleteUploadSessionRequest(StrictRequest):
    parts: list[UploadedPartRequest] = Field(default_factory=list, max_length=10_000)


class SignUploadPartRequest(StrictRequest):
    part_number: int = Field(
        ..., ge=1, le=10_000,
        validation_alias=AliasChoices("partNumber", "part_number"),
        serialization_alias="partNumber",
    )
    checksum_sha256: str = Field(
        ..., pattern=r"^[a-fA-F0-9]{64}$",
        validation_alias=AliasChoices("checksumSha256", "checksum_sha256"),
        serialization_alias="checksumSha256",
    )
    size_bytes: int = Field(
        ..., gt=0, le=64 * 1024 * 1024,
        validation_alias=AliasChoices("sizeBytes", "size_bytes"),
        serialization_alias="sizeBytes",
    )


class MutationRequest(StrictRequest):
    revision: int = Field(
        ..., ge=0,
        validation_alias=AliasChoices("revision", "expectedRevision", "expected_revision"),
        serialization_alias="revision",
    )


class ReplaceSourceRequest(CreateUploadSessionRequest):
    pass


class MarkSourcesReadyRequest(MutationRequest):
    pass


class GenerateVideoRequest(MutationRequest):
    idempotency_key: str = Field(
        ..., min_length=1, max_length=128,
        validation_alias=AliasChoices("idempotencyKey", "idempotency_key"),
        serialization_alias="idempotencyKey",
    )


class SourceErrorResponse(BaseModel):
    code: str
    message: str
    retryable: bool = False


class VideoSourceResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    folder_id: str = Field(serialization_alias="folderId")
    source_type: SourceType = Field(serialization_alias="sourceType")
    status: SourceStatus
    asset_id: str | None = Field(default=None, serialization_alias="assetId")
    display_name: str = Field(default="Source", serialization_alias="displayName")
    text_preview: str | None = Field(default=None, serialization_alias="textPreview")
    link_hostname: str | None = Field(default=None, serialization_alias="linkHostname")
    safe_metadata: dict[str, Any] = Field(default_factory=dict, serialization_alias="safeMetadata")
    preview_url: str | None = Field(default=None, serialization_alias="previewUrl")
    playback_url: str | None = Field(default=None, serialization_alias="playbackUrl")
    error: SourceErrorResponse | None = None
    error_code: str | None = Field(default=None, serialization_alias="errorCode")
    replacement_of_source_id: str | None = Field(default=None, serialization_alias="replacementOfSourceId")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class VideoSourceFolderResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    project_id: str = Field(serialization_alias="projectId")
    content_id: str = Field(serialization_alias="contentId")
    status: FolderStatus
    revision: int
    ready_revision: int | None = Field(default=None, serialization_alias="readyRevision")
    ready_at: datetime | None = Field(default=None, serialization_alias="readyAt")
    enqueue_status: EnqueueStatus = Field(serialization_alias="enqueueStatus")
    generation_request_id: str | None = Field(default=None, serialization_alias="generationRequestId")
    sources: list[VideoSourceResponse] = Field(default_factory=list)
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class UploadPartInstruction(BaseModel):
    part_number: int = Field(serialization_alias="partNumber")
    upload_url: str | None = Field(default=None, serialization_alias="uploadUrl")
    expires_at: datetime | None = Field(default=None, serialization_alias="expiresAt")
    size_bytes: int = Field(serialization_alias="sizeBytes")
    headers: dict[str, str] = Field(default_factory=dict)


class UploadSessionResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    session_id: str = Field(serialization_alias="sessionId")
    source_id: str = Field(serialization_alias="sourceId")
    strategy: Literal["proxy", "multipart"]
    upload_url: str | None = Field(default=None, serialization_alias="uploadUrl")
    parts: list[UploadPartInstruction] = Field(default_factory=list)
    expires_at: datetime = Field(serialization_alias="expiresAt")


class GenerationHandoffResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    folder_id: str = Field(serialization_alias="folderId")
    ready_revision: int = Field(serialization_alias="readyRevision")
    enqueue_status: EnqueueStatus = Field(serialization_alias="enqueueStatus")
    generation_request_id: str | None = Field(default=None, serialization_alias="generationRequestId")
    error: SourceErrorResponse | None = None


class GenerateVideoResultResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    folder: VideoSourceFolderResponse
    canonical_request_id: str | None = Field(default=None, serialization_alias="canonicalRequestId")

File: app/lib/data/models/video_source_intake.dart
import 'dart:typed_data';

enum VideoSourceType {
  binaryVideo('binary_video'),
  binaryImage('binary_image'),
  binaryAudio('binary_audio'),
  publicLink('public_link'),
  pastedText('pasted_text'),
  unknown('unknown');

  const VideoSourceType(this.id);
  final String id;

  static VideoSourceType parse(Object? value) => values.firstWhere(
    (entry) => entry.id == value?.toString(),
    orElse: () => unknown,
  );
}

enum VideoSourceStatus {
  pendingValidation('pending_validation'),
  processing('processing'),
  ready('ready'),
  metadataUnavailable('metadata_unavailable'),
  failed('failed'),
  replacementPending('replacement_pending'),
  superseded('superseded'),
  removed('removed'),
  orphanCleanupNeeded('orphan_cleanup_needed'),
  unknown('unknown');

  const VideoSourceStatus(this.id);
  final String id;

  bool get isActive => this != superseded && this != removed;
  bool get isBlocking => isActive && this != ready;
  bool get canRetry => switch (this) {
    failed || metadataUnavailable || orphanCleanupNeeded => true,
    _ => false,
  };

  static VideoSourceStatus parse(Object? value) => values.firstWhere(
    (entry) => entry.id == value?.toString(),
    orElse: () => unknown,
  );
}

enum VideoSourceFolderStatus {
  collecting('collecting'),
  ready('ready'),
  changedAfterReady('changed_after_ready'),
  archived('archived'),
  unknown('unknown');

  const VideoSourceFolderStatus(this.id);
  final String id;

  static VideoSourceFolderStatus parse(Object? value) => values.firstWhere(
    (entry) => entry.id == value?.toString(),
    orElse: () => unknown,
  );
}

enum VideoSourceEnqueueStatus {
  notRequested('not_requested'),
  enqueuePending('enqueue_pending'),
  enqueued('enqueued'),
  enqueueFailed('enqueue_failed'),
  unknown('unknown');

  const VideoSourceEnqueueStatus(this.id);
  final String id;

  static VideoSourceEnqueueStatus parse(Object? value) => values.firstWhere(
    (entry) => entry.id == value?.toString(),
    orElse: () => unknown,
  );
}

class VideoSourceSafeMetadata {
  const VideoSourceSafeMetadata({
    this.mimeType,
    this.sizeBytes,
    this.durationMs,
    this.width,
    this.height,
    this.characterCount,
    this.snippet,
    this.publicHost,
    this.publicTitle,
    this.previewUrl,
  });

  final String? mimeType;
  final int? sizeBytes;
  final int? durationMs;
  final int? width;
  final int? height;
  final int? characterCount;
  final String? snippet;
  final String? publicHost;
  final String? publicTitle;
  final String? previewUrl;

  factory VideoSourceSafeMetadata.fromJson(Map<String, dynamic> json) {
    return VideoSourceSafeMetadata(
      mimeType: _stringOrNull(json['mimeType'] ?? json['mime_type']),
      sizeBytes: _intOrNull(json['sizeBytes'] ?? json['size_bytes']),
      durationMs: _intOrNull(json['durationMs'] ?? json['duration_ms']),
      width: _intOrNull(json['width']),
      height: _intOrNull(json['height']),
      characterCount: _intOrNull(
        json['characterCount'] ?? json['character_count'] ?? json['char_count'],
      ),
      snippet: _stringOrNull(json['snippet']),
      publicHost: _stringOrNull(
        json['publicHost'] ?? json['public_host'] ?? json['hostname'],
      ),
      publicTitle: _stringOrNull(
        json['publicTitle'] ?? json['public_title'] ?? json['title'],
      ),
      previewUrl: _stringOrNull(json['previewUrl'] ?? json['preview_url']),
    );
  }
}

class VideoSource {
  const VideoSource({
    required this.id,
    required this.folderId,
    required this.type,
    required this.status,
    required this.displayName,
    this.assetId,
    this.safeMetadata = const VideoSourceSafeMetadata(),
    this.errorCode,
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String folderId;
  final VideoSourceType type;
  final VideoSourceStatus status;
  final String displayName;
  final String? assetId;
  final VideoSourceSafeMetadata safeMetadata;
  final String? errorCode;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  factory VideoSource.fromJson(Map<String, dynamic> json) {
    final metadata = Map<String, dynamic>.from(
      _map(json['safeMetadata'] ?? json['safe_metadata'] ?? json['metadata']),
    );
    metadata.putIfAbsent(
      'snippet',
      () => json['textPreview'] ?? json['text_preview'],
    );
    metadata.putIfAbsent(
      'publicHost',
      () => json['linkHostname'] ?? json['link_hostname'],
    );
    metadata.putIfAbsent(
      'previewUrl',
      () => json['previewUrl'] ?? json['preview_url'],
    );
    return VideoSource(
      id: _requiredString(json['id'], 'source id'),
      folderId: _requiredString(
        json['folderId'] ?? json['folder_id'],
        'folder id',
      ),
      type: VideoSourceType.parse(json['sourceType'] ?? json['source_type']),
      status: VideoSourceStatus.parse(json['status']),
      displayName:
          _stringOrNull(
            json['displayName'] ??
                json['display_name'] ??
                json['sourceLabel'] ??
                json['source_label'],
          ) ??
          'Source',
      assetId: _stringOrNull(json['assetId'] ?? json['asset_id']),
      safeMetadata: VideoSourceSafeMetadata.fromJson(metadata),
      errorCode: _stringOrNull(json['errorCode'] ?? json['error_code']),
      createdAt: _dateOrNull(json['createdAt'] ?? json['created_at']),
      updatedAt: _dateOrNull(json['updatedAt'] ?? json['updated_at']),
    );
  }
}

class VideoSourceFolder {
  const VideoSourceFolder({
    required this.id,
    required this.projectId,
    required this.contentId,
    required this.revision,
    required this.status,
    required this.enqueueStatus,
    required this.sources,
    this.readyRevision,
    this.canonicalRequestId,
    this.readyAt,
    this.updatedAt,
  });

  final String id;
  final String projectId;
  final String contentId;
  final int revision;
  final VideoSourceFolderStatus status;
  final int? readyRevision;
  final VideoSourceEnqueueStatus enqueueStatus;
  final String? canonicalRequestId;
  final List<VideoSource> sources;
  final DateTime? readyAt;
  final DateTime? updatedAt;

  List<VideoSource> get activeSources =>
      sources.where((source) => source.status.isActive).toList(growable: false);
  int get blockingSourceCount =>
      activeSources.where((source) => source.status.isBlocking).length;
  bool get canFinalize => activeSources.isNotEmpty && blockingSourceCount == 0;
  bool get isCurrentRevisionReady => readyRevision == revision;

  factory VideoSourceFolder.fromJson(Map<String, dynamic> json) {
    final rawSources = json['sources'] ?? json['items'];
    return VideoSourceFolder(
      id: _requiredString(json['id'] ?? json['folderId'], 'folder id'),
      projectId: _requiredString(
        json['projectId'] ?? json['project_id'],
        'project id',
      ),
      contentId: _requiredString(
        json['contentId'] ?? json['content_id'],
        'content id',
      ),
      revision: _intOrNull(json['revision']) ?? 0,
      status: VideoSourceFolderStatus.parse(json['status']),
      readyRevision: _intOrNull(
        json['readyRevision'] ?? json['ready_revision'],
      ),
      enqueueStatus: VideoSourceEnqueueStatus.parse(
        json['enqueueStatus'] ?? json['enqueue_status'] ?? 'not_requested',
      ),
      canonicalRequestId: _stringOrNull(
        json['canonicalRequestId'] ??
            json['canonical_request_id'] ??
            json['generationRequestId'] ??
            json['generation_request_id'],
      ),
      sources: rawSources is List
          ? rawSources
                .whereType<Map>()
                .map(
                  (entry) =>
                      VideoSource.fromJson(Map<String, dynamic>.from(entry)),
                )
                .toList(growable: false)
          : const <VideoSource>[],
      readyAt: _dateOrNull(json['readyAt'] ?? json['ready_at']),
      updatedAt: _dateOrNull(json['updatedAt'] ?? json['updated_at']),
    );
  }

  VideoSourceFolder copyWith({
    int? revision,
    VideoSourceFolderStatus? status,
    int? readyRevision,
    bool clearReadyRevision = false,
    VideoSourceEnqueueStatus? enqueueStatus,
    String? canonicalRequestId,
    bool clearCanonicalRequestId = false,
    List<VideoSource>? sources,
    DateTime? readyAt,
    DateTime? updatedAt,
  }) {
    return VideoSourceFolder(
      id: id,
      projectId: projectId,
      contentId: contentId,
      revision: revision ?? this.revision,
      status: status ?? this.status,
      readyRevision: clearReadyRevision
          ? null
          : (readyRevision ?? this.readyRevision),
      enqueueStatus: enqueueStatus ?? this.enqueueStatus,
      canonicalRequestId: clearCanonicalRequestId
          ? null
          : (canonicalRequestId ?? this.canonicalRequestId),
      sources: sources ?? this.sources,
      readyAt: readyAt ?? this.readyAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

class VideoSourceUploadFile {
  const VideoSourceUploadFile({
    required this.clientFileId,
    required this.fileName,
    required this.mimeType,
    required this.sizeBytes,
    this.bytes,
    this.path,
    this.readStream,
  });

  final String clientFileId;
  final String fileName;
  final String mimeType;
  final int sizeBytes;
  final Uint8List? bytes;
  final String? path;
  final Stream<List<int>>? readStream;

  Map<String, dynamic> toSessionJson() => {
    'clientFileId': clientFileId,
    'fileName': fileName,
    'mimeType': mimeType,
    'sizeBytes': sizeBytes,
  };
}

enum VideoSourceUploadTransport {
  proxy('proxy'),
  multipart('multipart'),
  presignedPut('presigned_put');

  const VideoSourceUploadTransport(this.id);
  final String id;

  static VideoSourceUploadTransport parse(Object? value) => values.firstWhere(
    (entry) => entry.id == value?.toString(),
    orElse: () => proxy,
  );
}

class VideoSourceUploadInstruction {
  const VideoSourceUploadInstruction({
    required this.clientFileId,
    required this.transport,
    required this.uploadTarget,
    this.method = 'PUT',
    this.headers = const <String, String>{},
  });

  final String clientFileId;
  final VideoSourceUploadTransport transport;
  final String uploadTarget;
  final String method;
  final Map<String, String> headers;

  factory VideoSourceUploadInstruction.fromJson(Map<String, dynamic> json) {
    _rejectProviderConfiguration(json);
    final target = _stringOrNull(
      json['uploadUrl'] ??
          json['upload_url'] ??
          json['uploadPath'] ??
          json['upload_path'],
    );
    if (target == null || target.isEmpty) {
      throw const FormatException('Opaque upload target is missing.');
    }
    final rawHeaders = _map(json['headers']);
    return VideoSourceUploadInstruction(
      clientFileId: _requiredString(
        json['clientFileId'] ?? json['client_file_id'],
        'client file id',
      ),
      transport: VideoSourceUploadTransport.parse(json['transport']),
      uploadTarget: target,
      method: _stringOrNull(json['method'])?.toUpperCase() ?? 'PUT',
      headers: rawHeaders.map(
        (key, value) => MapEntry(key, value?.toString() ?? ''),
      ),
    );
  }
}

class VideoSourceUploadSession {
  const VideoSourceUploadSession({
    required this.sessionId,
    required this.sourceId,
    required this.transport,
    required this.expiresAt,
    this.uploadUrl,
    this.parts = const <VideoSourceUploadPart>[],
  });

  final String sessionId;
  final String sourceId;
  final VideoSourceUploadTransport transport;
  final DateTime expiresAt;
  final String? uploadUrl;
  final List<VideoSourceUploadPart> parts;

  factory VideoSourceUploadSession.fromJson(Map<String, dynamic> json) {
    _rejectProviderConfiguration(json);
    final raw = json['parts'];
    return VideoSourceUploadSession(
      sessionId: _requiredString(
        json['sessionId'] ?? json['session_id'] ?? json['id'],
        'upload session id',
      ),
      sourceId: _requiredString(
        json['sourceId'] ?? json['source_id'],
        'source id',
      ),
      transport: VideoSourceUploadTransport.parse(
        json['strategy'] ?? json['transport'],
      ),
      expiresAt:
          _dateOrNull(json['expiresAt'] ?? json['expires_at']) ??
          DateTime.fromMillisecondsSinceEpoch(0, isUtc: true),
      uploadUrl: _stringOrNull(json['uploadUrl'] ?? json['upload_url']),
      parts: raw is List
          ? raw
                .whereType<Map>()
                .map(
                  (entry) => VideoSourceUploadPart.fromJson(
                    Map<String, dynamic>.from(entry),
                  ),
                )
                .toList(growable: false)
          : const <VideoSourceUploadPart>[],
    );
  }
}

class VideoSourceUploadPart {
  const VideoSourceUploadPart({
    required this.partNumber,
    required this.sizeBytes,
    this.uploadUrl,
    this.headers = const <String, String>{},
  });

  final int partNumber;
  final String? uploadUrl;
  final int sizeBytes;
  final Map<String, String> headers;

  factory VideoSourceUploadPart.fromJson(Map<String, dynamic> json) {
    _rejectProviderConfiguration(json);
    final rawHeaders = _map(json['headers']);
    return VideoSourceUploadPart(
      partNumber: _intOrNull(json['partNumber'] ?? json['part_number']) ?? 0,
      uploadUrl: _stringOrNull(json['uploadUrl'] ?? json['upload_url']),
      sizeBytes: _intOrNull(json['sizeBytes'] ?? json['size_bytes']) ?? 0,
      headers: rawHeaders.map(
        (key, value) => MapEntry(key, value?.toString() ?? ''),
      ),
    );
  }
}

class VideoSourceGenerateCommand {
  const VideoSourceGenerateCommand({
    required this.folderId,
    required this.revision,
    required this.idempotencyKey,
  });

  final String folderId;
  final int revision;
  final String idempotencyKey;

  Map<String, dynamic> toJson() => {
    'folderId': folderId,
    'revision': revision,
    'idempotencyKey': idempotencyKey,
  };
}

class VideoSourceGenerateResult {
  const VideoSourceGenerateResult({
    required this.folder,
    this.canonicalRequestId,
  });

  final VideoSourceFolder folder;
  final String? canonicalRequestId;

  factory VideoSourceGenerateResult.fromJson(Map<String, dynamic> json) {
    final folderJson = _map(json['folder']);
    return VideoSourceGenerateResult(
      folder: VideoSourceFolder.fromJson(folderJson),
      canonicalRequestId: _stringOrNull(
        json['canonicalRequestId'] ?? json['canonical_request_id'],
      ),
    );
  }
}

void _rejectProviderConfiguration(Map<String, dynamic> json) {
  const forbidden = <String>{
    'provider',
    'bucket',
    'bucketname',
    'objectkey',
    'storagekey',
    'bunnystoragekey',
    'bunnycdnhostname',
    'awsaccesskeyid',
    'awssecretaccesskey',
  };
  for (final key in json.keys) {
    final normalized = key.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]'), '');
    if (forbidden.contains(normalized)) {
      throw FormatException('Provider configuration is forbidden: $key');
    }
  }
}

Map<String, dynamic> _map(Object? value) {
  if (value is Map<String, dynamic>) return value;
  if (value is Map) return Map<String, dynamic>.from(value);
  return const <String, dynamic>{};
}

String _requiredString(Object? value, String label) {
  final parsed = _stringOrNull(value);
  if (parsed == null || parsed.isEmpty) {
    throw FormatException('Missing $label.');
  }
  return parsed;
}

String? _stringOrNull(Object? value) {
  if (value == null) return null;
  final parsed = value.toString().trim();
  return parsed.isEmpty ? null : parsed;
}

int? _intOrNull(Object? value) {
  if (value is int) return value;
  if (value is num) return value.toInt();
  return int.tryParse(value?.toString() ?? '');
}

DateTime? _dateOrNull(Object? value) {
  if (value is DateTime) return value;
  return DateTime.tryParse(value?.toString() ?? '');
}
